reps/2022-04-07-serialization.md

```diff
@@ -0,0 +1,314 @@
+
+Note that in real world we might simply wrap Msgpack to do serialization for common classes. Here we rebuild the wheel only to show how it works.
+
+```python
+class ListSerializer(RaySerializer):
+
+    def serialize(self, list_object: List[bytearray]) -> RaySerializationResult:
+        result = RaySerializationResult()
+        result.in_band_buffer = bytearray()
+        # Total element count.
+        result.in_band_buffer.extend(to_bytes(len(list_object))))
+        for element in list_object:
+            element_res = ray.serialize(element)
+            # in band buffer length
+            result.in_band_buffer.extend(to_bytes(len(element_res.in_band_buffer)))
+            # in band buffer
+            result.in_band_buffer.extend(element_res.in_band_buffer)
+            # append out of band buffer
+            result.out_of_band_buffers += element_res.out_of_band_buffers
+        return result
+
+    def deserialize(self, in_band_buffer: bytes,
+                    oob_iterator: Iterator[memoryview]) -> List[bytearray]:
+        result = []
+        pos = 0
+        # read element count
+        count, pos = get_int(in_band_buffer, pos)
+        for i in range(count):
+            # read element size
+            element_size, pos = get_int(in_band_buffer, pos)
+            # read element in-band data
+            element_in_band_data = in_band_buffer[pos: pos + element_size]
+            pos = pos + element_size
+            # nested serialization here
+            element = ray.deserialize(element_in_band_data, oob_iterator)
+            result.append(element)
+        return result
+```
+
+### Fallback Serializer
+
+By default, if no serializer is registered, use serializer in current Ray's version as the fallback. For example, pickle5 in Python, FST in Java.
+
+We'll register a fallback serializer to Ray. For example, pickle5 fallback serializer will be something like the following, all OOB buffers of pickle5 will be wrapped to Ray's OOB buffer.
+
+```python
+# Pseudo code
+class Pickle5FallbackSerializer(RaySerializer):
+    def serialize(self, instance) -> RaySerializationResult:
+        pickle_oob_buffers = []
+        in_band = pickle5.dumps(instance, buffer_callback=pickle_oob_buffers.append)
+        ray_oob_buffers = convert_to_ray_oob_buffers(pickle_oob_buffers)
+        return RaySerializationResult(in_band, ray_oob_buffers)
+
+    def deserialize(self, in_band_buffer: bytes,
+                    oob_iterator: Iterator[memoryview]):
+        return pickle5.loads(in_band_buffer, buffers=list(oob_iterator))
+```
+
+Nested serialization will still work in this case.
+When registering fallback serializer, we can provide a `register_callback` to receive notifications of serializer registration.
+
+```python
+def register_fallback_serializer(serializer, register_callback=None):
+    pass
+
+def register_serializer(class_name: str, class_type: type, serializer):
```

Comment:
Is it necessary to split class name & type? Can't one infer from the other?

Reply:
The class_name is a cross-lang ID, it's specified by users. We can't generate it from class_type.